### ISSUE TITLE
[Keras] Model does not learn at all - loss and val_acc remains constant - what other problems can be there?

### ISSUE THREAD

**Issue by neel04** [tensorflow/tensorflow#47586]
I am doing text classification and have made a dummy dataset to aid in debugging. The problem is that the model does not learn/converge. I am trying to predict one out of 20 labels and have tried most fixes on the net to no avail. 
```py
train_text = [['a'],['a'],['a']]
val_text = [['a'],['a'],['a']]
train_label = np.asarray([5,5,5])
val_label = np.asarray([5,5,5])

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

#Implementing a Transformer block as a layer

class TransformerBlock(layers.Layer):
    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
        super(TransformerBlock, self).__init__()
        self.att = layers.MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)
        self.ffn = keras.Sequential(
            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
        )
        self.layernorm1 = layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = layers.LayerNormalization(epsilon=1e-6)
        self.dropout1 = layers.Dropout(rate)
        self.dropout2 = layers.Dropout(rate)

    def call(self, inputs, training):
        attn_output = self.att(inputs, inputs)
        attn_output = self.dropout1(attn_output, training=training)
        out1 = self.layernorm1(inputs + attn_output)
        ffn_output = self.ffn(out1)
        ffn_output = self.dropout2(ffn_output, training=training)
        return self.layernorm2(out1 + ffn_output)


class TokenAndPositionEmbedding(layers.Layer):
    def __init__(self, maxlen, vocab_size, embed_dim):
        super(TokenAndPositionEmbedding, self).__init__()
        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=embed_dim)
        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=embed_dim)

    def call(self, x):
        maxlen = tf.shape(x)[-1]
        positions = tf.range(start=0, limit=maxlen, delta=1)
        positions = self.pos_emb(positions)
        x = self.token_emb(x)
        return x + positions

vocab_size = 40000  # Only consider the top 20k words :)
maxlen = 10  # for my task, ideally 500

train_label = tf.keras.utils.to_categorical(train_label, 20)
val_label = tf.keras.utils.to_categorical(val_label, 20)

tokenizer = tf.keras.preprocessing.text.Tokenizer(num_words=20000, lower=True, oov_token='<OOV>')
tokenizer.fit_on_texts(train_text)
train_sequences = tokenizer.texts_to_sequences(train_text)

val_sequences = tokenizer.texts_to_sequences(val_text)

train_text = tf.keras.preprocessing.sequence.pad_sequences(train_sequences, maxlen=maxlen, dtype='int32',
                                           padding='post', truncating='post')

val_text = tf.keras.preprocessing.sequence.pad_sequences(val_sequences, maxlen=maxlen, dtype='int32',
                                           padding='post', truncating='post')

from sklearn.preprocessing import normalize

train_text_n = normalize(train_text)
val_text_n = normalize(val_text)

data_set = tf.data.Dataset.from_tensor_slices((train_text_n , train_label))
data_set

from tensorflow.keras.regularizers import l2
 
embed_dim = 128  # Embedding size for each token
num_heads = 8  # Number of attention heads
ff_dim = 256  # Hidden layer size in feed forward network inside transformer
batch_size = 32
factor = 0.0001
 
inputs = layers.Input(shape=(maxlen,))

embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
x = embedding_layer(inputs)
transformer_block = TransformerBlock(embed_dim, num_heads, ff_dim)

x = transformer_block(x)

x = layers.GlobalAveragePooling1D()(x)

x = layers.Dense(160, activation="relu", kernel_initializer='glorot_uniform')(x)
x = layers.Dropout(0.15)(x)
x = layers.BatchNormalization()(x)

x = layers.Dense(90, activation="relu", kernel_initializer='glorot_uniform')(x)
x = layers.Dropout(0.1)(x)
x = layers.BatchNormalization()(x)

 
x = layers.Dense(40, activation="relu", kernel_initializer='glorot_uniform')(x)
x = layers.Dropout(0.1)(x)

outputs = layers.Dense(20)(x)
 
model = keras.Model(inputs=inputs, outputs=outputs)
 
adamopt = tf.keras.optimizers.Adam(learning_rate=1e-4)
 
 
model.compile(optimizer='adamopt', loss="categorical_crossentropy", metrics=["acc"])
 
model.summary()
 
history = model.fit(
    train_text_n, train_label, batch_size=batch_size, epochs=20, validation_data=(val_text_n, val_label), verbose=1)
```
Any Idea on how to do I fix this network? I am totally stumped after trying to fix it for 4 days :(

I posted this on Stack Overflow also, but seeing the lack of traffic, I doubt I would receive any response https://stackoverflow.com/questions/66471955/network-does-not-learn-what-are-the-other-things-that-can-go-wrong 

> There is a list of all the stuff I have tried on that link ^^^

**Comment by AdityaKane2001** [comment]
@neel04 
I tried your code in Colab, here's the [gist](https://colab.research.google.com/gist/AdityaKane2001/f2b018a594becbdc6bf2e411f3881d98/47586.ipynb).  As you mentioned in your SO issue,

> BUT the accuracy and val_accuracy remain constant.

In the above gist, the metrics keep changing, but as the dataset is quite small, they don't make any sense.  Is this the problem you were facing?
Also, if possible, could you please upload the logs and configuration details?

**Comment by neel04** [comment]
The problem was that even after removing the dropout and regularizations - the network does not overfit

**Comment by AdityaKane2001** [comment]
Ok, I found the issue
So the `train_text` has only one character, and thus the vectorized form is something as follows:
```
Train text: 
 [[2 0 0 0 0 0 0 0 0 0]
 [3 0 0 0 0 0 0 0 0 0]
 [4 0 0 0 0 0 0 0 0 0]
 [3 0 0 0 0 0 0 0 0 0]
 [2 0 0 0 0 0 0 0 0 0]
 [4 0 0 0 0 0 0 0 0 0]
 [2 0 0 0 0 0 0 0 0 0]
 [3 0 0 0 0 0 0 0 0 0]
 [4 0 0 0 0 0 0 0 0 0]]
```

Thus the normalized form becomes:
```
 [[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
```
Hence, the inputs are same for all samples, and thus the model is not training.

**Comment by neel04** [comment]
Good one, but it still does not overfit even if we don't normalize

**Comment by AdityaKane2001** [comment]
Yes, I tried that. The reason being, the function we are now trying to fit to is (for one sample):
```
output = np.zeros((20),dtype=np.float32)
output = np.put(sample, sample[0]+3 , 1 )
```

which I don't think is the best function to fit using a transformer :)

**Comment by AdityaKane2001** [comment]
I tried the sigmoid, but nothing changed

**Comment by neel04** [comment]
yeah, but I heard I have to use `categorical_crossentropy`

**Comment by AdityaKane2001** [comment]
That's correct, but the the function you are trying to fit is not possible, because it is not a curve that you can plot in some n-dimensional space.

**Comment by neel04** [comment]
no, I meant for the actual dataset I am using, should I use sigmoid?

**Comment by AdityaKane2001** [comment]
> no, I meant for the actual dataset I am using, should I use sigmoid?

yup

**Comment by AdityaKane2001** [comment]
Else you are just training a linear regression model 😅

**Comment by neel04** [comment]
yeah, I didn't see it that way. I didn't pick that up because in the examples I looked, they all used softmax. Well, it is training with sigmoid with a pretty small network (without normalization).

Thanx a lot Aditya!! :+1: :100: :rocket: 

Just one last thing - I heard that normalization can help in converging since there would be less interference of weights since the range is [1,40000]. Is there some other way/resource you recommend that I use to normalize my data?

**Comment by AdityaKane2001** [comment]
> yeah, I didn't see it that way. I didn't pick that up because in the examples I looked, they all used softmax. Well, it is training with sigmoid with a pretty small network (without normalization) so I expect the accuracy to scale up.

Ok, sorry about that. You must use softmax in your case, because each of your tokens have some probability associated with them. Sigmoid and softmax are close enough, just that sigmoid is for binary classification whereas softmax is for multilabel classification.

**Comment by AdityaKane2001** [comment]
About normalization: normalization works as it keeps the weights from overflow or underflow. Input data normalization can be done as you have done in the above code.

**Comment by neel04** [comment]
well, using softmax doesn't help at all :disappointed:

**Comment by AdityaKane2001** [comment]
Ok, I got a bit confused.
Softmax is used for `multi-label` classification : which means if output consists of more than one label per sample. 
Sigmoid is used for `multi-class` classification: which means if output has exactly one label out of many possible labels.
Most tutorials don't show the bigger picture: they show only binary  classification.

Thus, use sigmoid for your task. Sorry for the prior confusion. 😅 

Coming to normalization, make sure that you normalize along the correct axis. Secondly, note that dropout works effectively on extremely large networks (approx >1000 nodes per layer). This is just from my experience and not a mathematical fact. Also there's a difference between normalization and regularization: make sure you understand that.

**Comment by neel04** [comment]
Right now, situation is something like this:-
```py
Epoch 1/30
34/34 [==============================] - 13s 321ms/step - loss: 3.6056 - acc: 0.0571 - val_loss: 3.1510 - val_acc: 0.1003
Epoch 2/30
34/34 [==============================] - 11s 311ms/step - loss: 3.3849 - acc: 0.0598 - val_loss: 3.1064 - val_acc: 0.1003
Epoch 3/30
34/34 [==============================] - 11s 311ms/step - loss: 3.2823 - acc: 0.0480 - val_loss: 3.0968 - val_acc: 0.0418
Epoch 4/30
34/34 [==============================] - 11s 311ms/step - loss: 3.2642 - acc: 0.0633 - val_loss: 3.1047 - val_acc: 0.0362
Epoch 5/30
34/34 [==============================] - 11s 311ms/step - loss: 3.4409 - acc: 0.0542 - val_loss: 3.1140 - val_acc: 0.0306
```
it was working well when I initially used sigmoid, but after that, I made a lot of changes to try to get softmax working so now I don't know what exact things I used :(
Any possible insight?

Cheers,
Neel Gupta

**Comment by AdityaKane2001** [comment]
When using `one_hot` encoding, use `categorical_crossentropy`. Use sigmoid in last layer.
Also, please close this issue as we have now pinpointed the root cause.

**Comment by AdityaKane2001** [comment]
Could you get it to work?

**Comment by neel04** [comment]
no, it seems to be randomly guessing.

**Comment by AdityaKane2001** [comment]
Can you provide a small dataset so I can try on Colab?

**Comment by neel04** [comment]
[val_text.txt](https://github.com/tensorflow/tensorflow/files/6092019/val_text.txt)
[val_label.txt](https://github.com/tensorflow/tensorflow/files/6092020/val_label.txt)
[train_text.txt](https://github.com/tensorflow/tensorflow/files/6092021/train_text.txt)
[train_label.txt](https://github.com/tensorflow/tensorflow/files/6092022/train_label.txt)

I have attached 5 examples of each - you can try and see if it overfits.

**Comment by AdityaKane2001** [comment]
I don't know what's going on in these files 😅 . Can you share some data source that is easy to load?

**Comment by neel04** [comment]
In each file, there is a nested list that contains 5 other lists. they have been already converted to integers to facilitate mode training. you can simply copy the list, put it in a variable and train the model.
